front/lib/auth.ts

```diff
@@ -0,0 +1,86 @@
+import { App, User, DataSource } from "./models";
+import { InternalErrorWithStatusCode } from "./error";
+import { Result, Ok, Err } from "@app/lib/result";
+import { NextApiRequest, NextApiResponse } from "next";
+import { authOptions } from "@app/pages/api/auth/[...nextauth]";
+import { getServerSession } from "next-auth/next";
+
+type Role = "owner" | "read_only";
+
+export class Authenticator {
+  authUser: User;
+
+  constructor(authUser: User) {
+    this.authUser = authUser;
+  }
+
+  user() {
+    return this.authUser;
+  }
+
+  // There is no workspace yet but eventually we'll have a workspace model and this will be used to
+  // check if the user is a member of the workspace.
+  async roleFor(resourceOwner: User): Promise<Role> {
+    if (resourceOwner.id === this.authUser.id) {
+      return "owner";
+    } else {
+      return "read_only";
+    }
+  }
+
+  canReadApp(app: App): boolean {
+    switch (app.visibility) {
+      case "private":
+        return this.authUser.id === app.userId;
+      case "public":
+      case "unlisted":
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  canEditApp(app: App): boolean {
+    return this.authUser.id === app.userId;
+  }
+
+  canRunApp(app: App): boolean {
+    return this.authUser.id === app.userId;
```

Comment:
Yes for now, as otherwise the owner would see Runs appear in their interface.

In the future we'll have an `front` side `Run` object and we will allow running an app with your own credentials